docs: Add durable streaming migration guide, FAQ, and content enhancements (#1323)
vercel/workflow: docs/content/docs/faq/index.mdx

---
title: Frequently Asked Questions
description: Common questions about Workflow DevKit covering troubleshooting, migration, compatibility, and advanced usage.
type: guide
summary: Answers to common questions about streaming, debugging, migration, and WorkflowChatTransport.
---

## Getting unstuck

> ### Why does my stream stop when the user refreshes the page?
> 
> With standard streaming, the response is tied to a single HTTP connection. When the page reloads, the connection closes and the response is lost. Use [`WorkflowChatTransport`](/docs/api-reference/workflow-ai/workflow-chat-transport) to make the stream durable. The workflow keeps running on the server, and the client reconnects to the same run using the `runId`. See [Resumable Streams](/docs/ai/resumable-streams).
> 

Reviewer: We can't generally suggest using `WorkflowChatTransport` when the user might not even use the stream with an agent
```suggestion
```

> ### Why is my workflow step failing silently?
> 
> Open the Workflow Web UI locally and inspect the step execution trace:
> 
> ```bash
> npx workflow inspect runs --web
> ```
> 
> The step debugger shows the state of each step, including failures that do not surface in console logs. Click into a run to see the full step trace, retry attempts, and error details. See [Observability](/docs/observability).

Reviewer: Similarly not sure if I feel confident in this framing
```suggestion
```

### Why does `WorkflowChatTransport` ignore my custom body fields?

`WorkflowChatTransport` shapes its POST body differently than the default AI SDK transport. To add custom fields, use the `prepareSendMessagesRequest` hook:

```typescript
import { WorkflowChatTransport } from "@workflow/ai";

new WorkflowChatTransport({
  prepareSendMessagesRequest: async (config) => ({
    ...config,
    body: {
      ...config.body,
      customField: "value",
    },
  }),
})
```

See the [`WorkflowChatTransport` API Reference](/docs/api-reference/workflow-ai/workflow-chat-transport) for all options.

### Why does streaming need to live inside a step?

Workflow functions must be deterministic to support replay. Since streams bypass the [event log](/docs/how-it-works/event-sourcing) for performance, reading stream data in a workflow function would break determinism. By requiring all stream operations to happen in steps, the framework ensures consistent behavior across replays. See [Streaming](/docs/foundations/streaming#important-limitation).

### My reconnection endpoint returns an empty stream. What is wrong?

> Check that the `runId` in the reconnection URL matches the run you want to resume. You can inspect the run in the Web UI or CLI (the run does not need to be active to connect to a stream):

Reviewer: ```suggestion
Check that the `runId` in the reconnection URL matches the run you want to resume. You can inspect the run in the Web UI or CLI. The "streams" panel will tell you whether your stream content is populating correctly. The run does not need to be running in order to connect to a stream.
```

```bash

> npx workflow inspect runs

Reviewer: ```suggestion
npx workflow inspect runs --web
```

```

Also check that the `startIndex` query parameter is not set beyond the number of chunks the run has produced. If omitted, the stream starts from the beginning.

## Migration

### How do I migrate from ephemeral streaming to durable streaming?

The core changes are:

1. Move generation into a workflow function

> 2. Stream through `DurableAgent` and `getWritable()`

Reviewer: ```suggestion
2. Stream through `DurableAgent` and/or `getWritable()`
```

3. Return the `runId` in the response headers

> 4. Add a reconnection endpoint using `getRun()`

Reviewer: ```suggestion
4. Add a stream connection endpoint using `getRun()` and `run.getReadable()`
```

> 5. Use `WorkflowChatTransport` on the client

Reviewer: ```suggestion
5. Read from the endpoint from your client.
6. Optionally use `WorkflowChatTransport` when using AI SDK compatible streams for smooth reconnection
```

Yes. `DurableAgent` is designed for agentic workloads where the task outlives a single request-response cycle. It fits naturally with Workflow's execution model and gives you the same retries, observability, and local tooling. See [Building Durable AI Agents](/docs/ai).

## Advanced usage

### How does stream reconnection work after a network failure?

1. The client stores the `runId` from the initial workflow response header
2. If the stream is interrupted before receiving a "finish" chunk, `WorkflowChatTransport` automatically reconnects
3. The reconnection request includes the `startIndex` of the last chunk received
4. The server returns the stream from that position forward
5. The client continues rendering from where it left off

See [Resumable Streams](/docs/ai/resumable-streams) for a complete implementation.

### How do I handle user input mid-workflow?

Use [hooks](/docs/foundations/hooks). Workflow hooks pause execution and wait for external input before continuing. This fits use cases like confirmations, approvals, or user choices during a multi-step AI flow. See [Human in the Loop](/docs/ai/human-in-the-loop).

> ### What happens to in-progress runs when I redeploy?

Reviewer: ```suggestion
### What happens to in-progress runs when I make changes to my workflow?
```

> This depends on the [World](/docs/deploying) you are using. With the [Vercel World](/docs/deploying/world/vercel-world), runs continue to run on their original deployment, so deploying comes without risk. With the [Local World](/docs/deploying/world/local-world), the in-memory queue does not persist across server restarts, so in-progress runs will not resume.

Reviewer: ```suggestion
This depends on the [World](/docs/deploying) you are using. With the [Vercel World](/docs/deploying/world/vercel-world), runs continue to run on their original deployment, so deploying is risk-free.
With the [Local World](/docs/deploying/world/local-world), existing runs will try to run on the same code as new runs, which could lead to run failures if the code is not compatible. This is fine and can be ignored in local development.
```
